### ISSUE TITLE
Make instances of `error-class` a subclass of `tuple`, by making all concrete errors a subclass of a new class `error`, subclass of `tuple`

### ISSUE THREAD

**Issue by nomennescio** [factor/factor#2622]
Currently, the documentation for `ERROR:` incorrectly states:
```
The following two snippets are equivalent:
ERROR: invalid-values x y ;

TUPLE: invalid-values x y ;
: invalid-values ( x y -- * )
    \ invalid-values boa throw ;
```
which is clear from the definition of `ERROR:`:
`SYNTAX: ERROR: ... define-error-class ;`, where `define-error-class` sets the `"error-class"` word property on the tuple. Furthermore, if you query the class of an error object (created with e.g. `\ invalid-values boa`), you get `error-class`, a subclass of `tuple-class`. And although you can check if an object belongs to this class with `error-class?`, you can't create any method specialization on it. All such specializations are dispatched to either to the class of the tuple (e.g. `invalid-values`) or to `tuple`, with no possibility to discern between errors and regular tuples. Hence `call-next-method` is also not possible for tuples that are error classes.

I therefore propose to introduce a new class `error` (no significant nameclashes), and modify the definition of `ERROR:` so that it makes all created errors a subclass of `error`, a new subclass of `tuple`, representing the class of all errors.

This has no impact on existing code (other than a potential nameclash for code not in the Factor tree, which can be easily resolved)

**Comment by nomennescio** [comment]
One example of useful behavior to be able to specialize on would be `error.`. Currently we have:
```
M: tuple error. describe ;
```

whereas we then would be able to:
```
M: error error. ... ;
```
and represent an error in a more meaningful way as just a tuple with slots.

**Comment by mrjbq7** [comment]
Currently certain kernel-errors are arrays with special slots set according to type. And thrown from the VM. 

I’m unsure if we could make those “error” tuple instances easily.

**Comment by nomennescio** [comment]
Oh right. But I don't think I'm limiting those errors? I know any object can be thrown. I'm interested in those defined with `ERROR:`. Maybe it's possible to further generalize, and make a class hierarchy with different types of error classes, I'm not sure.

**Comment by erg** [comment]
Another thing that bothers me about ERROR: is the class name and the thrower word are the same. It shouldn't be like that imo.

**Comment by nomennescio** [comment]
Hmm, I did check the online docs for `error`, but when using the help Browser I see that name is also used by the `continuations` vocabulary as symbol to hold the most recently thrown error.

**Comment by nomennescio** [comment]
> Another thing that bothers me about ERROR: is the class name and the thrower word are the same. It shouldn't be like that imo.

It actually is biting me when writing my Codewars test framework to explicitly handle errors; I now need to create for each error:
```C: <some-error> some-error``` to be able to create error objects on the stack *without* immediately throwing them. I'm also not a big fan when encountering it in code, because it's not obvious at all a word immediately throws itself. There's not even a naming convention for them.

But that's a different matter altogether.

**Comment by nomennescio** [comment]
> Currently certain kernel-errors are arrays with special slots set according to type. And thrown from the VM. I’m unsure if we could make those “error” tuple instances easily.

Are you referring to this in error.cpp? : 
```
void factor_vm::general_error(vm_error_type error, cell arg1_, cell arg2_) {
...
cell error_object =
        allot_array_4(tag_fixnum(KERNEL_ERROR), tag_fixnum(error),
                      arg1.value(), arg2.value());
    ctx->push(error_object);
```

@mrjbq7 I don't think it's needed to make these instances of my proposed `error` class; these are not defined by `ERROR:`, hence are not tuples, so it won't make sense to overload on them. That wouldn't make my proposal less useful though.

If you would like to extend my proposal, you could change them into tuples or special objects and give them class `kernel-error`, and overload on them. I'm not sure if `error` and `kernel-error`  would need to be fitted in a kind of error class hierarchy, maybe they could.

**Comment by mrjbq7** [comment]
You can throw anything, strings, arrays, tuples, errors. 

Your comment leads to a discussion around whether you should only be able to throw errors, or have a class hierarchy of most but not all errors, or have generic dispatch on “error” tuples but then also dispatch against other errors that could be thrown like kernel-error?

My suggestion for code wars was a type like:

ERROR: user-code-error contained ;

And then when you have your recover block it could catch, wrap, and either re-throw or return a data stack with the user-code-error and then do something with the underlying error it contains. And it would work with all current types of errors, sort of by definition.

**Comment by nomennescio** [comment]
I'm aware you can throw anything. I was specifically looking at types defined with `ERROR:`, which are now indiscernible from regular tuples. I can work around that, but it's not elegant. I'm fine in expanding the discussion to all "errors", which makes it a more difficult discussion. Either way, types defined with `ERROR:` would need to be covered in that too. Do you have a more fundamental issue with making `ERROR:`s a subclass of some new class (whatever name you would choose for that) so that one can e.g. prettyprint `ERROR:`s, something that I'm missing?

**Comment by mrjbq7** [comment]
No, I was just saying we are going to have to work harder to make ERROR:
the only kinds of errors.

Also:

error class-of error-class?

That should be true for ERROR: defined tuples already.

**Comment by nomennescio** [comment]
OK, my intention was not to make all errors into an `ERROR:`, but I guess such a thing could be done. Maybe the confusion is in the Factor docs already: it treats `ERROR:` just "as if" it covers all errors:
```
Syntax sugar for defining errors:
 ERROR:
```
whereas it's more of a subclass of errors, because you don't need `ERROR:` at all to define errors, if you can throw any object. I think that's fine.

The issue might be more with naming that subclass `error`, as it might incorrectly suggest it contains all errors.

As for your `error class-of error-class?`, which `error` are you referring to?

**Comment by mrjbq7** [comment]
ERROR: foo ;

T{ foo } class-of error-class? t assert=

**Comment by nomennescio** [comment]
Yes, true, that's how I implement my current work-around:
```
M: tuple <somegeneric> dup class-of error-class? [ handle-error-case ] [ handle-tuple-case ] if ;
```
which is clumsy, ugly, and not fit for production code.

**Comment by nomennescio** [comment]
(Maybe the same case could be made for e.g. `SINGLETON:`. I wonder how it would help to avoid boilerplate code as can be seen in the prettyprinter)

**Comment by mrjbq7** [comment]
I am ALL in favor of reducing boilerplate and extraneous code!

**Comment by mrjbq7** [comment]
But your version can’t handle a non ERROR: thrown?

I assume it uses “is an error” to differentiate from “normal” test result.

**Comment by nomennescio** [comment]
> But your version can’t handle a non ERROR: thrown? I assume it uses “is an error” to differentiate from “normal” test result.

Are you talking about my proposal, or the code I'm using to handle exceptions?
As for my proposal, it's main use for me is currently prettyprinting in a suitable format. To that end, I just need to be able to discern regular tuples from `ERROR:`s. Any other object thrown cannot be detected to be an error by itself, but that's due to Factor's design.

If you're open to rethink that, and broaden the scope, I'm fine with that. It would be helpful if we could take a small step and make `ERROR:`s into their own class. If you want to postpone because of the bigger discussion, maybe we can do something on an experimental branch?

**Comment by mrjbq7** [comment]
Anything thrown (whether an error-class instance or string or array) is by definition an error. 

So if in your recover block you wrap the “error” with your tuple instance, then you can inspect that result and know an error occurred, whether it’s an error-class instance or not shouldn’t matter.  You know something was thrown.

**Comment by nomennescio** [comment]
> Anything thrown (whether an error-class instance or string or array) is by definition an error. So if in your recover block you wrap the “error” with your tuple instance, then you can inspect that result and know an error occurred, whether it’s an error-class instance or not shouldn’t matter. You know something was thrown.

That much is clear, but I'm talking about the general case where you can get any object, or a sequence of them, and want to use/iterate using (a) generic method(s).

**Comment by nomennescio** [comment]
Such an "error tuple" class could also be a simple predicate:
```
PREDICATE: user-error < tuple class-of error-class? ;
```
